### PR TITLE
Command Parser

### PR DESCRIPTION
**Pull request by SirRandoo**
Merged from [Utils](https://github.com/sirrandoo/toolkit-utils).

### CODE REVIEW

**Comment by hodlhodl1132** [comment]
1. When looking at the files changed, it appears to have the same code twice. Is this intended?

2. Also, if not too much trouble. Could you comment parts of the code so I can better understand whats intended with each part of the parsing?

**Comment by SirRandoo** [comment]
1. Artifact from when I pulled the latest changes
2. 👍

File: toolkitcore/Controllers/ChatCommandController.cs [hunk 1]
```diff
@@ -12,7 +12,172 @@ public static class ChatCommandController
...
+            // If any data is still in the buffer, we'll add it to the cache.
+            if (segment.Length > 0)
+            {
+                cache.Add(segment);
+            }
+
+            return cache.ToArray();
+        }
+
+        public static List<KeyValuePair<string, string>> ParseKeyed(string input, string prefix = "!")
```
Comment (hodlhodl1132) [inline toolkitcore/Controllers/ChatCommandController.cs]: Method never referenced

File: toolkitcore/Controllers/ChatCommandController.cs [hunk 2]
```diff
@@ -12,7 +12,172 @@ public static class ChatCommandController
...
+
+            return DefDatabase<ToolkitChatCommand>.AllDefsListForReading.FirstOrDefault(
+                c => c.commandText.EqualsIgnoreCase(baseCommand)
+            );
+        }
+
+        public static IEnumerable<string> Parse(string input, string prefix = "!")
+        {
+            // Strip the prefix from the input (if it exists)
+            if (input.ToLowerInvariant().StartsWith(prefix))
```
Comment (hodlhodl1132) [inline toolkitcore/Controllers/ChatCommandController.cs]: Command text input will never have '!'. ChatMessages without a '!' are never parsed to begin with.

**Pull request merged**